compiler/rustc_resolve/src/diagnostics.rs

```diff
@@ -950,17 +950,48 @@ impl<'a> Resolver<'a> {
         self.add_typo_suggestion(err, suggestion, ident.span);
 
         let import_suggestions =
-            self.lookup_import_candidates(ident, Namespace::MacroNS, parent_scope, |res| {
-                matches!(res, Res::Def(DefKind::Macro(MacroKind::Bang), _))
-            });
+            self.lookup_import_candidates(ident, Namespace::MacroNS, parent_scope, is_expected);
         show_candidates(err, None, &import_suggestions, false, true);
 
         if macro_kind == MacroKind::Derive && (ident.name == sym::Send || ident.name == sym::Sync) {
             let msg = format!("unsafe traits like `{}` should be implemented explicitly", ident);
             err.span_note(ident.span, &msg);
+            return;
         }
         if self.macro_names.contains(&ident.normalize_to_macros_2_0()) {
             err.help("have you added the `#[macro_use]` on the module/import?");
+            return;
+        }
+        for ns in [Namespace::MacroNS, Namespace::TypeNS, Namespace::ValueNS] {
+            if let Ok(binding) = self.early_resolve_ident_in_lexical_scope(
+                ident,
+                ScopeSet::All(ns, false),
+                &parent_scope,
+                false,
+                false,
+                ident.span,
+            ) {
+                let it_is = match binding.macro_kind() {
+                    Some(MacroKind::Bang) => "it is a function-like macro".to_string(),
+                    Some(kind) => format!("it is {} {}", kind.article(), kind.descr_expected()),
+                    None => format!(
+                        "it is not {} {}",
+                        macro_kind.article(),
+                        macro_kind.descr_expected()
+                    ),
```

Comment:
Instead of telling people what this _is not_, could we use `resolve_path` (or similar) to get what _it actually is_?

Reply:
Yes! I wanted to do that, but couldn't find a way to get that information. I figured that at this stage that info isn't available yet, but if that information exists at this point, we should do that!

Reply:
Looks like that information is already available in `binding.res()`. Apparently I didn't look hard enough. ^^'